base_agent/nsp_dialogue_manager.py

```diff
@@ -269,6 +294,16 @@ def get_logical_form(self, s: str, chat_as_list=False, ground_truth_actions: dic
             else:
                 d = self.model.parse(chat=s)
 
+        # Validate parse tree against grammar
+        is_valid_json = self.validate_parse_tree(d)
```

Comment:
how frequently do you expect this to be invalid? do we want to collect all the invalid parses - if so how?

Reply:
For ground truth, this should never be invalid, if it is, we need to fix ASAP. For parses from the semantic parser, it should not happen frequently. The model is generally good at producing parse trees that fit the grammar spec, even when they're wrong.

Agree, the wrong parses should be collected. I'll write to a separate file in datasets.

Reply:
Actually, may not need to write the failed parses to a separate file. Considering adding data checks to CI, then ground truth would always be up to date. Failed model parses aren't very actionable